instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py

```diff
@@ -0,0 +1,85 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional
+
+from aio_pika.abc import AbstractChannel, AbstractMessage
+
+from opentelemetry import context, trace
+from opentelemetry.semconv.trace import (
+    MessagingOperationValues,
+    SpanAttributes,
+)
+from opentelemetry.trace import Span, SpanKind
+
+from .version import __version__
+
+_DEFAULT_ATTRIBUTES = {SpanAttributes.MESSAGING_SYSTEM: 'rabbitmq'}
+
+
+class SpanBuilder:
+    TRACER_PROVIDER = None
+
+    def __init__(self):
+        self._attributes = _DEFAULT_ATTRIBUTES.copy()
+        self._operation: MessagingOperationValues = None
+        self._kind: SpanKind = None
+        self._destination: str = None
+
+    def set_as_producer(self):
+        self._kind = SpanKind.PRODUCER
+
+    def set_as_consumer(self):
+        self._kind = SpanKind.CONSUMER
+
+    def set_operation(self, operation: MessagingOperationValues):
+        self._operation = operation
+
+    def set_destination(self, destination: str):
+        self._destination = destination
+        self._attributes[SpanAttributes.MESSAGING_DESTINATION] = destination
+
+    def set_channel(self, channel: AbstractChannel):
+        url = channel.connection.connection.url
+        self._attributes.update({
+            SpanAttributes.NET_PEER_NAME: url.host,
+            SpanAttributes.NET_PEER_PORT: url.port
+        })
+
+    def set_message(self, message: AbstractMessage):
+        properties = message.properties
+        if properties.message_id:
+            self._attributes[SpanAttributes.MESSAGING_MESSAGE_ID] = properties.message_id
+        if properties.correlation_id:
+            self._attributes[SpanAttributes.MESSAGING_CONVERSATION_ID] = properties.correlation_id
+
+    def build(self) -> Optional[Span]:
+        if context.get_value('suppress_instrumentation') or context.get_value(context._SUPPRESS_INSTRUMENTATION_KEY):
+            return None
+        assert self._kind, 'kind must be configured.'
+        assert self._destination, 'destination must be configured.'
+        tracer = trace.get_tracer(__name__, __version__, self.TRACER_PROVIDER)
```

Comment:
I saw that in other instrumentations it is the same, see https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-pika/src/opentelemetry/instrumentation/pika/pika_instrumentor.py#L125

Reply:
In pika this happens per channel object, and only in case of manual instrumentation of an existing object. If you use the `PikaInstrumentor.instrument()` API it happens once. In case it happens per call for `build`, so every time you build a span.